add advanced-dependencies (refactorized)

epcim opened the pull request:
Fixes travis.ci build.
Adds ability to specify non local dependencies:
- git repo
- apt repo (see https://github.com/salt-formulas/salt-formula-*)
- yum repo
- spm repo
- ppa repo

Obviously allow complex cross formula testing without requirement to have all dependencies locally etc.
For kitchen-salt build it uses git formula with dependencies defined in formula metadata.yml.
Add option to specify additional init commands in kitchen yml.

(tested with git and apt)
Feedback welcome.

Rejected:
- fetch dependencies on host (not in remote guest) --> for git repos it's possible for apt/yum not. Spm possibly yes. But for "git" at least we already have a solution, (I think vendor_path) - is a place which is  uploaded to test instance. Anyway as the code to fetch git formulas with dependencies already exists, I will share a script to auto/fetch dependencies in: https://github.com/salt-formulas/salt-formulas/tree/master/deploy/scripts. (together with some other nice scripts)

ek9 commented on lib/kitchen/provisioner/install.erb:
```diff
@@ -8,11 +9,12 @@ salt_apt_repo_key = config[:salt_apt_repo_key]
 
 <<-INSTALL
-sh -c '
+bash -c '
```
Why not stick using to `sh`? `bash` is not installed on all environments, while `sh` will be symlinked to a `sh` compatible shell.

ek9 commented on lib/kitchen/provisioner/install.erb:
```diff
@@ -72,8 +72,72 @@ else
 
-#{install_chef}
+echo "-----> Installing other dependencies..."
+YUM_CMD="$(which yum)"
```
the use of `which` should be avoided. Info on why and what to use instead on https://stackoverflow.com/a/677212

ek9 commented on lib/kitchen/provisioner/install.erb:
```diff
@@ -72,8 +72,72 @@ else
+YUM_CMD="$(which yum)"
+APT_CMD="$(which apt-get)"
+APK_CMD="$(which apk)"
+if [[ -z "$(which git)" ]]; then
```
It would be nice to have the scripts as separate `.sh` files. We can lint them to make sure they are `sh` compatible and will work well in different environments. Having one bootstrap script should be sufficient.

Othewise, for git command etc, why not depend on ruby? It feels kind of weird to mix up bash and ruby like this.

epcim committed:
Advanced dependencies

- ability to specify git/spm/apt/yum dependencies
- refactorized to keep shell files out of ruby
- shellcheck verification for .sh files
- updates for travis yml

Martin819 committed:
fix duplicate name in inherited dependecies, typo for shellcheck pass

epcim committed:
Merge pull request #4 from Martin819/master

fix duplicate name in inherited dependecies, typo for shellcheck pass
